controllers/imagerepository_controller.go

```diff
@@ -577,32 +590,216 @@ func (r *ImageRepositoryReconciler) EnsureSecret(ctx context.Context, imageRepos
+
+	serviceAccount := &corev1.ServiceAccount{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: buildPipelineServiceAccountName, Namespace: namespace}, serviceAccount)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		log.Error(err, "Failed to read pipeline service account", "ServiceAccountName", buildPipelineServiceAccountName, "NamespaceName", namespace, l.Action, l.ActionView)
+		return err
+	}
+
+	linksModified := false
+
+	// Check for duplicates for the secret and remove them
+	pushSecrets := []corev1.ObjectReference{}
+	foundSecret := false
+	for _, credentialSecret := range serviceAccount.Secrets {
+		if credentialSecret.Name == secretName {
+			if !foundSecret {
+				pushSecrets = append(pushSecrets, credentialSecret)
+				foundSecret = true
+			} else {
+				linksModified = true
 			}
+		} else {
+			pushSecrets = append(pushSecrets, credentialSecret)
+		}
+	}
+	serviceAccount.Secrets = pushSecrets
+
+	// Remove secret from pull secrets list
+	imagePullSecrets := []corev1.LocalObjectReference{}
+	for _, pullSecret := range serviceAccount.ImagePullSecrets {
+		// don't break and search for duplicities
+		if pullSecret.Name == secretName {
+			linksModified = true
+			continue
+		}
+		imagePullSecrets = append(imagePullSecrets, pullSecret)
+	}
+	serviceAccount.ImagePullSecrets = imagePullSecrets
+
+	if linksModified {
+		if err := r.Client.Update(ctx, serviceAccount); err != nil {
+			log.Error(err, "Failed to update pipeline service account", "ServiceAccountName", buildPipelineServiceAccountName, "NamespaceName", namespace, l.Action, l.ActionUpdate)
+			return err
+		}
+		log.Info("Cleaned up secret links in pipeline service account", "SecretName", secretName, "ServiceAccountName", buildPipelineServiceAccountName, l.Action, l.ActionUpdate)
+	}
+	return nil
+}
+
+// VerifyAndFixSecretsLinking ensures that the given secret is linked to the provided service account, and also removes duplicated link for the secret.
+// when secret doesn't exist unlink it from service account
+func (r *ImageRepositoryReconciler) VerifyAndFixSecretsLinking(ctx context.Context, imageRepository *imagerepositoryv1alpha1.ImageRepository) error {
+	log := ctrllog.FromContext(ctx)
+
+	secretName := imageRepository.Status.Credentials.PushSecretName
+
+	// check secret existence and if secret doesn't exist unlink it from service account
+	// users can recreate it by using regenerate-token
+	secret := &corev1.Secret{}
+	secretKey := types.NamespacedName{Namespace: imageRepository.Namespace, Name: secretName}
+	secretExists := true
+	if err := r.Client.Get(ctx, secretKey, secret); err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "Failed to get secret", "SecretName", secretName, l.Action, l.ActionView)
+			return err
+		}
+
+		secretExists = false
+		log.Info("Secret doesn't exist, will unlink secret from service account", "SecretName", secretName)
+		log.Info("To recreate secret use regenerate-token")
```

Comment:
This log is not visible to the user, so makes no sense.

Reply:
I know it isn't visible to the user, but it is good for debugging

keeping it

Reply:
For debugging we can relay on `Secret doesn't exist, will unlink secret from service account`, cannot we?

Reply:
ok removing